test/unittests/validation_test.cpp

```diff
@@ -113,4 +113,202 @@ TEST(validation, br_table_default_invalid_label_index)
         "0061736d01000000010401600000030201000a14011200024041000e080001000100010001020b0b");
 
     EXPECT_THROW_MESSAGE(parse(wasm), validation_error, "invalid label index");
-}
\ No newline at end of file
+}
+
+TEST(validation, stack_underflow)
+{
+    /* wat2wasm --no-check
+    (func (param i32 i32) (result i32)
+      get_local 0
+      get_local 1
+      i32.add
+      i32.add
+    )
+    */
+    const auto wasm =
+        from_hex("0061736d0100000001070160027f7f017f030201000a0a010800200020016a6a0b");
+    EXPECT_THROW_MESSAGE(parse(wasm), validation_error, "stack underflow");
+}
+
+TEST(validation, stack_underflow_in_block)
+{
+    /* wat2wasm --no-check
+    (func
+      i32.const 2
+      (block
+        drop
+      )
+    )
+    */
+    const auto wasm = from_hex("0061736d01000000010401600000030201000a0a010800410202401a0b0b");
+    EXPECT_THROW_MESSAGE(parse(wasm), validation_error, "stack underflow");
+}
+
+TEST(validation, stack_underflow_in_block_with_result)
```

Comment:
This one doesn't have underflow, maybe should be renamed?
```suggestion
TEST(validation, stack_with_block_result)
```